data/src/main/java/com/microsoft/azure/kusto/data/AadAuthenticationHelper.java

```diff
@@ -3,300 +3,282 @@
+             *  CompletableFuture so that this will be an async method and let the SDK user wait for the value per their
+             *  needs, though I doubt others would find it worthwhile. Else, I can unblock the thread with sleep(exponentialBackoff)
+             *  as I proposed with AAD_ACCESS_TOKEN_PROVIDER, but I doubt others would find that worthwhile either.
+             */
+            result = future.get(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+        } catch (MalformedURLException e) {
+            throw new DataClientException(scope, ERROR_INVALID_AUTHORITY_URL, e);
+        } catch (ExecutionException | TimeoutException e) {
+            throw new DataServiceException(scope, ERROR_ACQUIRING_APPLICATION_ACCESS_TOKEN, e);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new DataServiceException(scope, ERROR_ACQUIRING_APPLICATION_ACCESS_TOKEN, e);
         }
 
         if (result == null) {
-            throw new DataServiceException(clusterUrl, "acquireAadApplicationAccessToken got 'null' authentication result");
+            throw new DataServiceException(scope, "acquireWithAadApplicationKey got 'null' authentication result");
         }
         return result;
     }
 
-    private AuthenticationResult acquireAccessTokenUsingDeviceCodeFlow() throws Exception {
-        AuthenticationContext context = null;
-        AuthenticationResult result = null;
-        ExecutorService service = null;
+    protected IAuthenticationResult acquireWithAadApplicationClientCertificate() throws DataServiceException, DataClientException {
+        IAuthenticationResult result;
         try {
-            service = Executors.newSingleThreadExecutor();
-            context = new AuthenticationContext(aadAuthorityUri, true, service);
-            Future<DeviceCode> future = context.acquireDeviceCode(CLIENT_ID, clusterUrl, null);
-            DeviceCode deviceCode = future.get();
-            System.out.println(deviceCode.getMessage());
-            if (Desktop.isDesktopSupported()) {
-                Desktop.getDesktop().browse(new URI(deviceCode.getVerificationUrl()));
-            }
-            result = waitAndAcquireTokenByDeviceCode(deviceCode, context);
-
-
-        } finally {
-            if (service != null) {
-                service.shutdown();
-            }
-        }
-        if (result == null) {
-            throw new ServiceUnavailableException("authentication result was null");
+            lastConfidentialClientApplication = ConfidentialClientApplication.builder(applicationClientId, clientCertificate).authority(aadAuthorityUrl).validateAuthority(false).build();
+            CompletableFuture<IAuthenticationResult> future = lastConfidentialClientApplication.acquireToken(ClientCredentialParameters.builder(scopes).build());
+            result = future.get(TIMEOUT_MS, TimeUnit.MILLISECONDS);
```

Comment:
Check if MSAL has built in timeouts (would be better then leaving the request open while you timeout)

Reply:
You're absolutely right - MSAL provides both connect and read timeouts, but they have reasonable defaults (10s and 15s, respectively).
Do you think we should override those defaults?

Considering the above, I think my future's timeout is too high. I'll lower it to 20s. We should never reach that (as we should at least get a failure response by then), but just in case...

Reply:
I'd trust their defaults unless proven otherwise.
They may also provide a retry option (MS Graph does for instance)

Reply:
I tend to want resiliency at every level. Indeed their code has a timeout, but I prefer every async call have its own failsafe timeout just in case something takes too long within MSAL other than the part they added a timeout to. It shouldn't hurt anything, and simply makes our code more resilient.

It's not a big deal, so I can take it out. I just think it's better being deliberate with timeouts at every level.